fs/exfat/namei.c

```diff
@@ -98,36 +98,29 @@ static int exfat_d_hash(const struct dentry *dentry, struct qstr *qstr)
 	const unsigned char *name;
 	unsigned int len;
 	unsigned long hash;
+	int i, charlen;
 
 	name = qstr->name;
 	len = exfat_striptail_len(qstr);
 
 	hash = __exfat_init_name_hash(dentry);
 
 	if (EXFAT_SB(sb)->options.utf8) {
-		int i, unilen;
-		unsigned short *uniname;
-
-		uniname	= kcalloc(MAX_NAME_LENGTH + 3,
-				sizeof(unsigned short), GFP_KERNEL);
-		if (!uniname)
-			return -ENOMEM;
-
-		unilen = utf8s_to_utf16s(name, strlen(name), UTF16_HOST_ENDIAN,
-				(wchar_t *)uniname, MAX_NAME_LENGTH + 2);
-		if (unilen < 0) {
-			kfree(uniname);
-			return unilen;
-		}
+		int j;
+		wchar_t utf16[8];
 
-		for (i = 0; i < unilen; i++)
-			hash = partial_name_hash(exfat_toupper(sb,
-					uniname[i]), hash);
-		kfree(uniname);
+		for (i = 0; i < len; i += charlen) {
+			charlen = utf8s_to_utf16s(&name[i], 4,
+				UTF16_HOST_ENDIAN, utf16, 8);
+			if (charlen <= 0 || charlen > 8)
+				return charlen;
+			for (j = 0; j < charlen; j++)
+				hash = partial_name_hash(exfat_toupper(sb,
+					utf16[j]), hash);
+		}
```

Comment:
Above code does not work. Function `utf8s_to_utf16s()` expects valid UTF-8 string of specified size in second argument (in your case `4`). So if `name` would contain string with two code points, one encoded as 1 byte long and second as 4 bytes long, above function call returns error (incomplete 3 byte sequence from second code point).

I would suggest to use `utf8_to_utf32()` and then one Unicode code point (UTF-32) in `unicode_t` type convert to UTF-16. Below is simple implementation. I have not tested it, so use it just a reference how it could look like.
```c
unicode_t u;
int size;
for (i = 0; i < len; i += charlen) {
    charlen = utf8_to_utf32(&name[i], len - i, &u);
    if (charlen < 0)
        return charlen;
    /* Convert to UTF-16: code points above U+FFFF are encoded as surrogate pairs */
    if (u > 0xFFFF) {
        /* exfat_toupper() works only for code points up to the U+FFFF */
        hash = partial_name_hash(((u - 0x10000) >> 10) + 0xD800, hash);
        hash = partial_name_hash(((u - 0x10000) & 0x3FF) + 0xDC00, hash);
    } else {
        hash = partial_name_hash(exfat_toupper(sb, (unsigned short)u), hash);
    }
}
```